File: ios/ios/PlatformContext_iOS.m
//
//  PlatformContext_iOS.m
//  ios
//
//  Created by Sahil Ishar on 3/10/16.
//  Copyright © 2016 Kevin Galligan. All rights reserved.
//

#import "PlatformContext_iOS.h"
#import "co/touchlab/droidconandroid/presenter/ConferenceDayHolder.h"
#import "co/touchlab/droidconandroid/presenter/ScheduleBlockHour.h"
#import "co/touchlab/droidconandroid/data/UserAccount.h"
#import "co/touchlab/droidconandroid/presenter/AppManager.h"
#import "ios-Swift.h"

@implementation PlatformContext_iOS


-(id)init {
    if ( self = [super init] ) {
        self.iosContext = [DCPAppManager getContext];
    }
    return self;
}

- (AndroidContentContext *)appContext {
    return self.iosContext;
}

- (NSString *)storageDir
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsDirectory = [paths objectAtIndex:0];
    
    return documentsDirectory;
}

- (NSMutableArray *)convertIOSObjectArrayToArray:(IOSObjectArray *)objArray
{
    NSMutableArray *array = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < objArray.length; i++) {
        DCPScheduleBlockHour *holder = [objArray objectAtIndex:i];
        [array addObject:holder];
    }

    return array;
}

- (NSMutableArray *)getHourBlocksArray
{
    IOSObjectArray *objArray;
    NSMutableArray *array = [[NSMutableArray alloc] init];
    
    int index = self.isDayTwo ? 1 : 0;
    
    if (self.conferenceDays.count > index)
    {
        DCPConferenceDayHolder *daySchedule = [self.conferenceDays objectAtIndex:index];
        objArray = daySchedule->hourHolders_;
        [array addObjectsFromArray:[self convertIOSObjectArrayToArray:objArray]];
    }
    
    if (self.dateFormatter == nil) {
        self.dateFormatter = [[NSDateFormatter alloc] init];
        [self.dateFormatter setDateFormat:@"hh:mma"];
    }
    
    return array;
}

- (NSArray *)getSpeakersArrayFromEvent:(DCDEvent *)event
{
    return [PlatformContext_iOS javaListToNSArray:event->speakerList_];
}

+ (NSArray *)javaListToNSArray:(id<JavaUtilList>)list
{
    NSMutableArray *array = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < [list size]; i++) {
        [array addObject:[list getWithInt:(jint)i]];
    }
    
    return array;
}

- (NSString *)formatSpeakerStringFromArray:(NSArray *)array
{
    NSMutableArray *speakerNames = [[NSMutableArray alloc] init];
    
    for (DCDEventSpeaker *speaker in array) {
        [speakerNames addObject:speaker->userAccount_->name_];
    }
    
    return [speakerNames componentsJoinedByString:@", "];
}

- (NSString *)getEventTimeFromStart:(NSString *)startTime andEnd:(NSString *)endTime
{
    int startLoc = 7;
    int endLoc = 7;
    int startLocCutoff = 0;
    
    NSString *startFirstDigit = [startTime substringWithRange:NSMakeRange(startTime.length-7, 1)];
    NSString *endFirstDigit = [endTime substringWithRange:NSMakeRange(endTime.length-7, 1)];
    
    if ([startFirstDigit isEqualToString:@"0"]) {
        startLoc = 6;
    }
    if ([endFirstDigit isEqualToString:@"0"]) {
        endLoc = 6;
    }
    
    NSString *startAmPm = [startTime substringWithRange:NSMakeRange(startTime.length-2, 2)];
    NSString *endAmPm = [endTime substringWithRange:NSMakeRange(endTime.length-2, 2)];
    if ([startAmPm isEqualToString:endAmPm]) {
        startLocCutoff = 2;
    }
    
    NSString *startTimeStr = [startTime substringWithRange:NSMakeRange(startTime.length-startLoc, startLoc-startLocCutoff)];
    NSString *endTimeStr = [endTime substringWithRange:NSMakeRange(endTime.length-endLoc, endLoc)];
    
    return [NSString stringWithFormat:@"%@ - %@", startTimeStr, endTimeStr];
}

- (NSString *)getFullEventTime:(DCDEvent *)event
{
    NSString *day = [[event getStartFormatted] substringWithRange:NSMakeRange(0, 10)];

    if (self.timeFormatter == nil) {
        self.timeFormatter = [[NSDateFormatter alloc] init];
    }
    
    [self.timeFormatter setDateFormat:@"MM/dd/yyyy"];
    NSDate* myDate = [self.timeFormatter dateFromString:day];
    [self.timeFormatter setDateFormat:@"MMMM dd"];
    NSString *dayString = [self.timeFormatter stringFromDate:myDate];
    
    return [NSString stringWithFormat:@"%@, %@", dayString, [self getEventTimeFromStart:[event getStartFormatted] andEnd:[event getEndFormatted]]];
}

- (void)updateTableData
{
    [self.hourBlocks removeAllObjects];
    [self.hourBlocks addObjectsFromArray:[self getHourBlocksArray]];
}

- (void)loadCallbackWithDCPConferenceDayHolderArray:(IOSObjectArray *)conferenceDayHolders
{
    self.hourBlocks = [[NSMutableArray alloc] init];
    
    self.conferenceDays = (NSArray *)conferenceDayHolders;
    [self updateTableData];
    [self.reloadDelegate reloadTableView];
    
}

#pragma Table View - Delegate

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    int index = self.isDayTwo ? 1 : 0;
    
    if(self.conferenceDays == nil || self.conferenceDays.count <= index)
        return 0;

    DCPConferenceDayHolder *daySchedule = [self.conferenceDays objectAtIndex:index];
    return daySchedule->hourHolders_->size_;

}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    static NSString *cellIdentifier = @"cellIdentifier";
    
    EventListCell *cell = (EventListCell*)[tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    
    DCPScheduleBlockHour *hourHolder = [self.hourBlocks objectAtIndex:[indexPath row]];
//    NSArray *scheduleBlocks = (NSArray *)hourHolder->scheduleBlocks_;
    
    NSObject *eventObj = hourHolder->scheduleBlock_;
    if ([eventObj isKindOfClass:[DCDEvent class]]) {
        DCDEvent *event = (DCDEvent *)eventObj;
        NSArray *speakers = [self getSpeakersArrayFromEvent:event];
        [cell.titleLabel setText: [event->name_ stringByReplacingOccurrencesOfString:@"Android" withString:@"[Sad Puppy]"]];
        [cell.speakerNamesLabel setText:[self formatSpeakerStringFromArray:speakers]];
        [cell.timeLabel setText:hourHolder->hourStringDisplay_];
        
        if([event isRsvped] && ![event isPast])
        {
            cell.rsvpView.hidden = true;
        }
        else{
            cell.rsvpView.hidden = false;
        }
//        [cell.textLabel setTextColor:[UIColor blackColor]];
//        [cell.detailTextLabel setTextColor:[UIColor colorWithRed:(151/255.0) green:(151/255.0) blue:(151/255.0) alpha:1.0]];
//        [cell setBackgroundColor:[UIColor whiteColor]];
//        [cell setUserInteractionEnabled:YES];
    } else {
        DCDBlock *event = (DCDBlock *)eventObj;
        
        [cell.titleLabel setText:event->name_];
        [cell.speakerNamesLabel setText:[self getEventTimeFromStart:[event getStartFormatted] andEnd:[event getEndFormatted]]];
        [cell.timeLabel setText:@" "];
        cell.rsvpView.hidden = true;
                
//        [cell.textLabel setTextColor:[UIColor colorWithRed:(87/255.0) green:(125/255.0) blue:(140/255.0) alpha:1.0]];
//        [cell.detailTextLabel setTextColor:[UIColor colorWithRed:(87/255.0) green:(125/255.0) blue:(140/255.0) alpha:1.0]];
//        [cell setBackgroundColor:[UIColor lightGrayColor]];
//        [cell setUserInteractionEnabled:NO];
    }
    
    cell.layer.opaque = YES;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
    
//    DCPConferenceHourHolder *hourHolder = [self.hourBlocks objectAtIndex:[indexPath section]];
//    NSArray *scheduleBlocks = (NSArray *)hourHolder->scheduleBlocks_;
    
    NSObject *eventObj = ((DCPScheduleBlockHour*)[self.hourBlocks objectAtIndex:[indexPath row]])->scheduleBlock_;
    if ([eventObj isKindOfClass:[DCDEvent class]]) {
        DCDEvent *event = (DCDEvent *)eventObj;
        [self.reloadDelegate showEventDetailViewWithEvent:event andIndex:[indexPath row]];
    }
}

@end

File: ios/ios/PlatformContext_iOS.swift
//
//  PlatformContext_iOS.swift
//  ios
//
//  Created by Kachi Nwaobasi on 7/11/17.
//  Copyright © 2017 Kevin Galligan. All rights reserved.
//

import Foundation
import JRE
import UIKit

protocol PlatformContext_iOSDelegate : class {
    func reloadTableView()
    func showEventDetailView(with timeBlock: DDATTimeBlock, andIndex index: Int)
}

class PlatformContext_iOS : NSObject {

    var isDayTwo: Bool = false
    var dateFormatter: DateFormatter!
    var timeFormatter: DateFormatter!
    var hourBlocks: [DPRESHourBlock]!
    var conferenceDays: [DPRESDaySchedule]?
    
    weak var reloadDelegate: PlatformContext_iOSDelegate?

    lazy var storageDir: String = {
        let paths = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true)
        return paths[0]
    }()
    
    //TODO: What is this? Looks like this doesn't need to be here
    var hourBlocksArray : [DPRESHourBlock] {
        var array = [DPRESHourBlock]()
        
        let index = self.isDayTwo ? 1 : 0
        if let days = conferenceDays, days.count > index, let objectArray = days[index].getHourHolders() {
            array.append(contentsOf: JavaUtils.convertiOSObjectArrayToArray(objArray: objectArray) as! [DPRESHourBlock])
        }
        if dateFormatter == nil {
            dateFormatter = DateFormatter()
            dateFormatter.dateFormat="hh:mma"
        }
        return array
    }
    
    func getSpeakersArray(from networkEvent: DDATEvent) -> [Any] {
        return JavaUtils.javaList(toList: networkEvent.getSpeakerList())
    }
    
    func getEventTime(startTime: NSString, andEnd endTime: NSString) -> String {
        var startLoc = 7
        var endLoc = 7
        var startLocCutoff = 0
        
        let startFirstDigit = (startTime as NSString).substring(with: NSMakeRange((startTime as NSString).length - 7, 1))
        let endFirstDigit = (endTime as NSString).substring(with: NSMakeRange((endTime as NSString).length - 7, 1))
        
        if startFirstDigit == "0" {
            startLoc = 6
        }
        if endFirstDigit == "0" {
            endLoc = 6
        }
        
        let startAmPm = startTime.substring(with: NSMakeRange(startTime.length - 2, 2))
        let endAmPm = endTime.substring(with: NSMakeRange(endTime.length - 2, 2))
        if startAmPm == endAmPm {
            startLocCutoff = 2
        }
        
        let startTimeStr = startTime.substring(with: NSMakeRange(startTime.length-startLoc, startLoc-startLocCutoff))
        let endTimeStr = endTime.substring(with: NSMakeRange(endTime.length - endLoc, endLoc))
        
        return "\(startTimeStr) - \(endTimeStr)"
    }
    
    func updateTableData() {
        hourBlocks.removeAll()
        hourBlocks.append(contentsOf: hourBlocksArray)
    }
}

extension PlatformContext_iOS : UITableViewDataSource {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let index = isDayTwo ? 1 : 0
        guard let days = self.conferenceDays, days.count > index else {
            return 0
        }
        let daySchedule = days[index]
        return Int(daySchedule.getHourHolders().length())
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellIdentifier", for: indexPath) as! ScheduleListCell
        
        let hourHolder = hourBlocks[indexPath.row]
        let eventObj = hourHolder.getTime()
        
        if let event = eventObj as? DDATEvent {
            cell.titleLabel.text = event.getName().replacingOccurrences(of: "Android", with: "[Sad Puppy]")
            cell.speakerNamesLabel.text = event.allSpeakersString()
        } else if let event = eventObj as? DDATBlock {
            cell.titleLabel.text = event.getName()
            cell.speakerNamesLabel.text = ""
        }
        cell.timeLabel.text = hourHolder.getHourStringDisplay().lowercased()
        cell.startOfBlock = indexPath.row == 0 || hourBlocks[indexPath.row - 1].getTime().getStartLong() != eventObj!.getStartLong()
        cell.layer.isOpaque = true
        return cell
    }
    
}

extension PlatformContext_iOS : UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, didHighlightRowAt indexPath: IndexPath) {
        tableView.cellForRow(at: indexPath)?.isHighlighted = true
    }
    
    func tableView(_ tableView: UITableView, didUnhighlightRowAt indexPath: IndexPath) {
        tableView.cellForRow(at: indexPath)?.isHighlighted = false
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
        
        let eventObj = hourBlocks[indexPath.row].getTime()
        if let networkEvent = eventObj {
            
            reloadDelegate?.showEventDetailView(with: networkEvent, andIndex: indexPath.row)
        }
    }
    
}

extension PlatformContext_iOS : DPRESScheduleDataViewModel_Host {

    func loadCallback(withDPRESDayScheduleArray daySchedules: IOSObjectArray!) {
        hourBlocks = [DPRESHourBlock]()
        conferenceDays = JavaUtils.convertiOSObjectArrayToArray(objArray: daySchedules) as? [DPRESDaySchedule]
        updateTableData()
        reloadDelegate?.reloadTableView()
    }
}

